python/lsst/rubintv/production/watchers.py

```diff
@@ -0,0 +1,226 @@
+# This file is part of rubintv_production.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import json
+import logging
+from glob import glob
+from time import sleep
+
+import lsst.daf.butler as dafButler
+from lsst.utils.iteration import ensure_iterable
+
+from .uploaders import Heartbeater
+
+from .utils import (raiseIf,
+                    writeDataIdFile,
+                    getGlobPatternForDataProduct)
+
+_LOG = logging.getLogger(__name__)
+
+
+class FileWatcher:
+    """A file-system based watcher, looking for outputs from a ButlerWatcher.
+
+    The ButlerWatcher polls the repo looking for dataIds, and writes them out
+    to a file, to be found by a File watcher.
+
+    Many of these can be instantiated per-location.
+
+    It is worth noting that the ``dataProduct`` to watch for need not be an
+    official dafButler dataset type. We can use FileWatchers to signal to any
+    downstream processing that something is finished and ready for consumption.
+
+    Uploads a heartbeat to the bucket every ``HEARTBEAT_PERIOD`` seconds if
+    ``heartbeatChannelName`` is specified.
+    """
+    cadence = 1  # in seconds
+
+    # upload heartbeat every n seconds
+    HEARTBEAT_UPLOAD_PERIOD = 30
+    # consider service 'dead' if this time exceeded between heartbeats
+    HEARTBEAT_FLATLINE_PERIOD = 120
+
+    def __init__(self, *, locationConfig, dataProduct, heartbeatChannelName='', doRaise=False):
+        self.locationConfig = locationConfig
+        self.dataProduct = dataProduct
+        self.doRaise = doRaise
+        self.log = _LOG.getChild("fileWatcher")
+        self.doHeartbeat = False
+        if heartbeatChannelName:
+            self.doHeartbeat = True
+            self.heartbeatChannelName = heartbeatChannelName  # currently unused
+            self.heartbeater = Heartbeater(heartbeatChannelName,
+                                           self.locationConfig.bucketName,
+                                           self.HEARTBEAT_UPLOAD_PERIOD,
+                                           self.HEARTBEAT_FLATLINE_PERIOD)
```

Comment:
Since you don't seem to use `self.heartbeatChannelName` after this at all, it might be better to replace this with:
```
if heartbeatChannelName:
    self.heartbeater = Heartbeater(...)
else:
    self.heartbeater = None
```
and then use `if self.heartbeater is not None` instead of `self.doHeatbeat`.

I do think it'd be good to at least set `self.heartbeater` (and `self.heartbeatChannel`) to `None` in an `else` branch here.  Not having attributes at all sometimes doesn't feel right to me, though I don't have a concrete explanation for why `None` is better than leaving them undefined.

Reply:
Yeah, totally agree, not really sure why I did it like that tbh.